TypeCobol/Compiler/CodeModel/SymbolTable.cs

```diff
@@ -148,62 +148,51 @@ public SymbolTable(SymbolTable enclosing, Scope current)
         public IDictionary<string, List<DataDefinition>> DataEntries =
             new Dictionary<string, List<DataDefinition>>(StringComparer.OrdinalIgnoreCase);
 
+        //Dictionary for Type data entries
+        public IDictionary<string, List<DataDefinition>> DataTypeEntries =
+            new Dictionary<string, List<DataDefinition>>(StringComparer.OrdinalIgnoreCase);
+
         internal void AddVariable([NotNull] DataDefinition symbol)
         {
             // TODO: generate a name for FILLERs and anonymous data to be referenced by in the symbol table
             if (symbol.Name == null)
             {
                 return;
             }
-            Add(DataEntries, symbol);
+            //If symbol is a part of a type add to DataTypeEntries
+            if (!symbol.IsPartOfATypeDef)
+            {
+                Add(DataEntries, symbol);
+            }
+            else
+            {
+                //Add DataDefinition for symbol in the DataTypeEntries dictionary
+                AddTypeVariableFromDataDefinition(symbol);
+            }
         }
 
-        internal void AddVariable(string name, DataDefinition data)
+        void AddTypeVariableFromDataDefinition([NotNull] DataDefinition data)
         {
-            string key = name;
-            List<DataDefinition> found;
-            bool present = DataEntries.TryGetValue(key, out found);
-            if (!present)
+            //Types are declared in the Declarations SymbolTable
+            var table = GetTableFromScope(Scope.Declarations);
+
+            if (table == null)
+            {
+                return;
+            }
+            //Add symbol to the dictionary
+            Add(table.DataTypeEntries, data);
+            //Add indexes as they need special treatment
+            foreach (var child in data.GetChildren<IndexDefinition>())
```

Comment:
IndexDefinition are added to SymbolTable directly by ProgramClassBuilder.
However, current ProgramClassBuilder doesn't add them to SymbolTable if it's under a TypeDef.


So remove "//Add indexes as they need special treatment" and make ProgramClassBuilder add Index into SymbolTable (check for comment "//If index is inside a Typedef do not add to symboltable" in ProgramClassBuilder.).